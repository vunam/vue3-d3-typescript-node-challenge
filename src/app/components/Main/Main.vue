<template>
  <div class="Main-container">
    <div v-if="state.nodeList.length" >
      <h1 class="Main-title">ABN AMRO Code Challenge</h1>
      <TreeChart :data="parentChildData" />
    </div>
     <div v-else>
      Loading...
    </div>
  </div>
 
</template>

<style lang="css" scope>
    .Main-container {
      flex-grow: 1;
      padding: 20px;
    }
    .Main-title {
      margin-top: 0;
    }
</style>

<script lang="ts">
import { defineComponent, inject } from "vue"
import { stratify } from "d3-hierarchy";
import { StateProps, ApiNode } from "../../types"
import TreeChart from "../TreeChart/TreeChart";

export default defineComponent({
  inject: ['state', 'setState'],
  async mounted() {
    const response = await fetch('http://localhost:8888/all');
    const result = await response.json();
    this.setState('nodeList', result);
  },
  components: {TreeChart},
  computed: {
    parentChildData() {
      if (!this.state.nodeList.length) {
        return [];
      }

      const root = stratify()
        .id((d: ApiNode) => d.name)
        .parentId((d: ApiNode) => d.parent)(
        this.state.nodeList)

      return root;
    }
  }
})
</script>