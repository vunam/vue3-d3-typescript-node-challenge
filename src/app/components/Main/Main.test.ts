import { shallowMount } from '@vue/test-utils';
import Main from './Main.vue';

describe('Main.vue', () => {
	beforeAll(() => {
		global.fetch = jest.fn().mockImplementation(() => ({
			json: () => [ 'some-data' ]
		}));
	});
	afterAll(() => {
		(global.fetch as jest.MockedFunction<any>).mockClear();
		delete global.fetch;
	});

	it('matches snapshot when theres no data', () => {
		const wrapper = shallowMount(Main, { propsData: {} });
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('fetch to be called on mount', () => {
		const wrapper = shallowMount(Main);
		expect(global.fetch).toHaveBeenCalled();
	});

	it('matches snapshot when data is received', () => {
		const wrapper = shallowMount(Main, {
			data() {
				return { data: [ 'some-data' ] };
			}
		});
		expect(wrapper.html()).toMatchSnapshot();
	});
});
