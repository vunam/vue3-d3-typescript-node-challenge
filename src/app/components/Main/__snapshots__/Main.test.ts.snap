// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main.vue matches snapshot when data is received 1`] = `"<div class=\\"Main-container\\"><h1 class=\\"Main-title\\">ABN AMRO Code Challenge</h1> [\\"some-data\\"]</div>"`;

exports[`Main.vue matches snapshot when theres no data 1`] = `"<div> Loading... </div>"`;
